src/features/definitions.ts

```diff
@@ -1,26 +1,47 @@
 import * as vscode from 'vscode'
 import {InstanceManager} from '../extension'
 import {serverProtocol} from 'ensime-client'
+import logapi = require("loglevel")
+
+const log = logapi.getLogger('ensime.definitions')
 
 export function definitionsProvider(instanceManager: InstanceManager) : vscode.DefinitionProvider {
-    return {
+
+    const provider : vscode.DefinitionProvider = {
+
        provideDefinition(document: vscode.TextDocument, 
             position: vscode.Position,
             token: vscode.CancellationToken) : Thenable<vscode.Definition> {
+           log.debug('provideDefinition called for ', document.fileName)
            const instance = instanceManager.instanceOfFile(document.fileName)
+           if (!instance) {
+               return new Promise((r, _) => r([]))
+           }
            return instance.api.getSymbolAtPoint(document.fileName, document.offsetAt(position)).then((symbol) => {
                const pos = symbol.declPos
                // TODO: Gotta be a smarter way in ts to do this, at least localise in client code
                if(pos.typehint == 'OffsetSourcePosition') {
                    const offsetPos = <serverProtocol.OffsetSourcePosition> pos
-                   // TODO: How to we resolve a character offset into a row,col position before we have the buffer?
                    const uri = vscode.Uri.file(offsetPos.file)
-                   return new vscode.Location(uri, new vscode.Position(offsetPos.row-1, offsetPos.col))
+                   const openedDoc = vscode.workspace.textDocuments.find(doc => doc.fileName == offsetPos.file);
+                   if (openedDoc) {
+                       const pos = openedDoc.positionAt(offsetPos.offset)
+                       return new vscode.Location(uri, pos);
+                   } else {
+                       return vscode.workspace.openTextDocument(uri).then(definitionDoc => {
+                           vscode.window.showTextDocument(definitionDoc)
```

Comment:
I was thinking in these lines too, but isn't this going to open up a document everytime you just hover? You have the definitions hover thingy too so it happens way before an actual click…